src/rbtree.rs

```diff
@@ -7,34 +8,86 @@
 
 //! Intrusive red-black tree.
 
-use crate::Bound::{self, Excluded, Included, Unbounded};
-use crate::IntrusivePointer;
-use crate::{Adapter, KeyAdapter};
 use core::borrow::Borrow;
 use core::cell::Cell;
 use core::cmp::Ordering;
 use core::fmt;
+use core::hint;
 use core::mem;
-use core::ptr;
+use core::ptr::NonNull;
+
+use crate::Bound::{self, Excluded, Included, Unbounded};
+
+use crate::link_ops::{self, DefaultLinkOps};
+use crate::linked_list::LinkedListOps;
+use crate::pointer_ops::PointerOps;
+use crate::singly_linked_list::SinglyLinkedListOps;
+use crate::xor_linked_list::XorLinkedListOps;
+use crate::Adapter;
+use crate::KeyAdapter;
+
+// =============================================================================
+// RBTreeOps
+// =============================================================================
+
+/// The color of a red-black tree node.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+#[allow(missing_docs)]
+pub enum Color {
+    Red,
+    Black,
+}
+
+/// Link operations for `RBTree`.
+pub unsafe trait RBTreeOps: link_ops::LinkOps {
+    /// Returns the left child of `ptr`.
+    fn left(&self, ptr: Self::LinkPtr) -> Option<Self::LinkPtr>;
```

Comment:
Operations that only read a value also need to be `unsafe` since we are passing arbitrary pointers into it.